overlord/snapstate/cookies.go

```diff
@@ -0,0 +1,137 @@
+	"github.com/snapcore/snapd/strutil"
+)
+
+// GenerateCookies creates snap cookies for snaps that are missing them (may be the case for snaps installed
+// before the feature of running snapctl outside of hooks was introduced, leading to a warning
+// from snap-confine).
+// It is the caller's responsibility to lock state before calling this function.
+func (m *SnapManager) GenerateCookies(st *state.State) error {
+	var snapNames map[string]*json.RawMessage
+	if err := st.Get("snaps", &snapNames); err != nil && err != state.ErrNoState {
+		return err
+	}
+
+	var snapCookies map[string]string
+	if err := st.Get("snap-cookies", &snapCookies); err != nil {
+		if err != state.ErrNoState {
+			return fmt.Errorf("cannot get snap cookies: %v", err)
+		}
+		snapCookies = make(map[string]string)
+	}
+
+	snapWithCookies := make(map[string]bool)
+	for _, snap := range snapCookies {
+		snapWithCookies[snap] = true
+	}
+
+	var changed bool
+	for snap := range snapNames {
+		if _, ok := snapWithCookies[snap]; !ok {
+			cookieID, err := createCookieFile(snap)
+			if err != nil {
+				return err
+			}
+
+			snapCookies[cookieID] = snap
+			changed = true
+		}
+	}
+
+	if changed {
+		st.Set("snap-cookies", &snapCookies)
+	}
+	return nil
+}
+
+func (m *SnapManager) createSnapCookie(st *state.State, snapName string) error {
+	var snapCookies map[string]string
+	if err := st.Get("snap-cookies", &snapCookies); err != nil {
+		if err != state.ErrNoState {
+			return fmt.Errorf("cannot get snap cookies: %v", err)
+		}
+		snapCookies = make(map[string]string)
+	}
+
+	// make sure we don't create cookie if it already exists
+	for _, snap := range snapCookies {
+		if snapName == snap {
+			return nil
+		}
+	}
+
+	cookieID, err := createCookieFile(snapName)
+	if err != nil {
+		return err
+	}
+
+	snapCookies[cookieID] = snapName
+	st.Set("snap-cookies", &snapCookies)
+	return nil
+}
+
+func (m *SnapManager) removeSnapCookie(st *state.State, snapName string) error {
+	var snapCookies map[string]string
+	err := st.Get("snap-cookies", &snapCookies)
+	if err != nil {
+		if err != state.ErrNoState {
+			return fmt.Errorf("cannot get snap cookies: %v", err)
+		}
+		// no cookies in the state
+		if err := os.Remove(filepath.Join(dirs.SnapCookieDir, fmt.Sprintf("snap.%s", snapName))); err != nil && !os.IsNotExist(err) {
+			return err
+		}
+		return nil
+	}
+
+	for cookieID, snap := range snapCookies {
+		if snapName == snap {
+			if err := os.Remove(filepath.Join(dirs.SnapCookieDir, fmt.Sprintf("snap.%s", snapName))); err != nil && !os.IsNotExist(err) {
+				return err
+			}
+			delete(snapCookies, cookieID)
+			st.Set("snap-cookies", snapCookies)
+			return nil
+		}
+	}
+	return nil
+}
+
+func createCookieFile(snapName string) (cookieID string, err error) {
+	cookieID = strutil.MakeRandomString(44)
+	path := filepath.Join(dirs.SnapCookieDir, fmt.Sprintf("snap.%s", snapName))
+	err = osutil.AtomicWriteFile(path, []byte(cookieID), 0600, 0)
```

Comment:
what happens if the file already exist?  (there's a chance that disk and state get out of sync)

Reply:
It's ok to overwrite cookie file with a new cookie, but that shouldn't happen since createSnapCookie method (which cals createCookie) will prevent that (cookie won't be generate if we have it in state). I've modified GenerateCookies method (named SyncCookies now) to use EnsureDirState on startup to make sure state and disk are in sync.

Reply:
ok, it will overwrite.  I made the remark because  we could crash between creating the file and setting the state in createSnapCookie